e2e/common_test.go

```diff
@@ -49,3 +53,76 @@ func RunCmd(cmd *exec.Cmd) (stdoutBytes []byte, stderrBytes []byte) {
 
 	return stdout.Bytes(), stderr.Bytes()
 }
+
+type forwardedPort struct {
+	local    int
+	remote   int
+	protocol string
+}
+
+func startForwardingTenantAPI(name, namespace string, localPort int) (net.Listener, error) {
+	address, err := net.ResolveIPAddr("", "127.0.0.1")
+	port := forwardedPort{
+		local:    localPort,
+		remote:   6443,
+		protocol: "tcp",
+	}
+	if err != nil {
+		return nil, err
+	}
+	listener, err := net.ListenTCP(
+		port.protocol,
+		&net.TCPAddr{
+			IP:   address.IP,
+			Zone: address.Zone,
+			Port: port.local,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	go waitForConnection(name, namespace, listener, port)
+
+	return listener, nil
+}
+
+func waitForConnection(name, namespace string, listener net.Listener, port forwardedPort) {
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			glog.Errorln("error accepting connection:", err)
+			return
+		}
+		stream, err := virtClient.VirtualMachineInstance(namespace).PortForward(name, port.remote, port.protocol)
+		if err != nil {
+			glog.Errorf("can't access vmi %s/%s: %v", namespace, name, err)
+			return
+		}
+		go handleConnection(conn, stream.AsConn(), port)
+	}
+}
+
+// handleConnection copies data between the local connection and the stream to
+// the remote server.
+func handleConnection(local, remote net.Conn, port forwardedPort) {
+	errs := make(chan error)
```

Comment:
make this a buffer of 2 so it's non-blocking (two go routines need to be able to write to this channel)